### PR TITLE
RJSF enhancements / simplifications

### LINKED ISSUE #365: In generated rjsf, exclude empty objects, and hide unnecessary fields
Based on UX feedback, and also common sense, we should not be showing empty cards on the form, and also unnecessary fields should be excluded as appropriate.

### PR DESCRIPTION
**Pull request by cisaacstern**
closes #365

### CODE REVIEW

**Comment by cisaacstern** [comment]
Current status of this work:

- Any task instances which _do not require_ user input are not omitted entirely from the generated RJSF (instead of being included as an empty card).
- The `persist_text.filename` argument is excluded from the form, since that should also never be set by users.
- An [`ADVANCED_KWS` constant is added]([95874d1](https://github.com/wildlife-dynamics/ecoscope-workflows/pull/364/commits/95874d112994c9261d91db252c35e7cb24a220d9)) which when used i.e.

**Comment by walljcg** [comment]
Thanks Charles. What about dataclasses. E.g., :

**Review by atmorling — COMMENTED** [review]
The only other thing I'd throw out there is what if we extended FieldInfo?
something like:
```python
from pydantic import FieldInfo

class AdvancedField(FieldInfo):
    def __init__(self, *args, **kwargs):
        kwargs["exclude"] = True
        super().__init__(*args, **kwargs)
```

Then we could have
```python
@task
def apply_reloc_coord_filter(
    df: AnyGeoDataFrame,
    min_x: Annotated[
        float,
        AdvancedField(default=-180.0),
    ] = -180.0,
    max_x: Annotated[float, AdvancedField(default=180.0)] = 180.0,
    min_y: Annotated[float, AdvancedField(default=-90.0)] = -90.0,
    max_y: Annotated[float, AdvancedField(default=90.0)] = 90.0,
    filter_point_coords: Annotated[
        list[Coordinate], AdvancedField(default=[Coordinate(x=0.0, y=0.0)])
    ] = [Coordinate(x=0.0, y=0.0)],
) -> AnyGeoDataFrame:
    import geopandas  # type: ignore[import-untyped]
``` 
Which IMO reads a lot nicer, but there's maybe reasons I haven't considered for not doing it this way.

**Comment by walljcg** [comment]
Where are the Column Color Options 1..n coming from? Not seeing those in either the field defaults or spec.yaml? 

Screenshot 2024-11-01 at 1 37 00 PM

**Comment by cisaacstern** [comment]
> What about dataclasses. E.g., : 

I believe Annotated can be used here? 

```python
@dataclass(frozen=True)
class Grouper:
    index_name: IndexName
    display_name: Annotated[str | SkipJsonSchema[None] = None, Field(**ADVANCED)]
```

**Comment by cisaacstern** [comment]
> if we extended FieldInfo?
something like:

Yeah that does look better, good call.

I think it's easier to just make it this a function? 

```python
from pydantic import Field

def AdvancedField(...):
    return Field(..., exclude=True)
```

File: src/ecoscope-workflows-core/ecoscope_workflows_core/tasks/io/_persist.py [hunk 1]
```diff
@@ -15,15 +14,15 @@
...
     filename: Annotated[
-        str | SkipJsonSchema[None],
+        str | None,
         Field(
             description="""\
             Optional filename to persist text to within the `root_path`.
             If not provided, a filename will be generated based on a hash of the text content.
-            """
+            """,
+            exclude=True,
```
[the remark this answers is not in the document]
Comment (cisaacstern) [reply src/ecoscope-workflows-core/ecoscope_workflows_core/tasks/io/_persist.py:25]: Well so i guess there is the question of ADVANCED vs. "this should _never_ be visible in the frontend" and this is the latter.

[the remark this answers is not in the document]
Comment (atmorling) [reply src/ecoscope-workflows-core/ecoscope_workflows_core/tasks/io/_persist.py:25]: Ah, yep. Makes sense.

**Comment by cisaacstern** [comment]
@walljcg I've merged your first pass of changes into this PR via #370. When you get back to this, you can continue on your same branch (or a new one) and we can merge any further changes into this PR as a subsequent round.

Some notes:

- Advanced fields on nested models now work. The recommended approach is to annotate the field in the same style as task function parameters, e.g. `Annotated[int, AdvancedField(default=1)] = 1` for an int field.
- I realized that a requirement for all AdvancedFields is that they must have a default (which can be None). Otherwise, they will be unset when we call the task function (because they are not set by a user). So I have made that a requirement in the AdvancedField constructor, which will throw ValueError if default is not provided (as either the first positional or a named kw argument). 
- The Advanced field default _will_ be used for all workflows, so if all workflows cannot run with a given default, then it can't be an AdvancedField. As such, e.g., I think `relocs_columns` is disqualified because its value differs between workflows? Unless there is a way around that, possibly by using the broadest possible set, and having some logic in the task function to handle per-dataset variation.
- generally all of your advanced params made sense to me, expect for the Groupers display_value, see https://github.com/wildlife-dynamics/ecoscope-workflows/pull/370#discussion_r1826350585
- there are a few test failures here which I can address Monday but they should not interfere with carrying on with your work in the interim (they seem isolated to the async tests for some unclear reason, but do not appear to be a generalized issue, and shouldn't be related to rjsf generation)

After you've given a full pass on AdvancedFields, noting here that we'll also want to look at:

- task groups
- descriptions (of groups and parameters, and possibly tasks themselves with the addition of a `@task(description=...)`)
- variable header font weight of nested objects in the same card in rjsf

**Comment by walljcg** [comment]
@cisaacstern Think on Monday we should discuss how to handle the fields where there are multiple types of input e.g., get_color for layers. This I would say is not an advanced field and we want to expose it but the rsjf is wonky.

**Comment by cisaacstern** [comment]
Agreed re: union types. The easiest thing is probably if we could just support one type for now at the RJSF layer, even if the Python layer can theoretically support one of a set of types. For get_color specifically, it would be helpful for someone with better knowledge of how this value is consumed to give us some guidance as to whether or not this is realistic in this case; ie can we just accept str here?

Good point re: defaults in the spec. This would alleviate our challenge with relocs_columns, perhaps, and probably others as well. One challenge here (not of implementation so much as documentation) is that this would only be possibly for fields where the inputs are deserializable from yaml/json-supported types; ie can't set defaults for fields that require a geodataframe, etc. Perhaps there is a way to encode this as a check into the compilation process, I have seen something related to this in the pydantic docs 🤔

**Comment by cisaacstern** [comment]
This failure appears to be the root of all the ci failures following merge of #372

https://github.com/wildlife-dynamics/ecoscope-workflows/actions/runs/11673120505/job/32503208498?pr=364

and surfacing the error for visibility

```
FAILED tests/test_app.py::test_generate_nested_params - fastapi.exceptions.ResponseValidationError: 2 validation errors:
  {'type': 'string_type', 'loc': ('response', 'colormap_traj_speed', 'colormap'), 'msg': 'Input should be a valid string', 'input': ['#1a9850', '#91cf60', '#d9ef8b', '#fee08b', '#fc8d59', '#d73027']}
  {'type': 'string_type', 'loc': ('response', 'colormap_traj_night', 'colormap'), 'msg': 'Input should be a valid string', 'input': ['#292965', '#e7a553']}
```

And the place these parameters are being attempted to be passed

https://github.com/wildlife-dynamics/ecoscope-workflows/blob/0abc3f4da1831868d6defa88c09536786e634fc8/examples/subject-tracking/test-cases.yaml#L59-L60

https://github.com/wildlife-dynamics/ecoscope-workflows/blob/0abc3f4da1831868d6defa88c09536786e634fc8/examples/subject-tracking/test-cases.yaml#L92

So I'll note that, not that I think about it, this is actually _expected_ (even desired, perhaps). Because `SkipJsonSchema` affects not just what goes in the RJSF, but (and these are deliberately linked), what types are _allowed_ to be POSTed as part of the `formData` generated from the RJSF.

For this specific case, I _think_ once I fix #373, we can still use these custom color maps, if we set that as explicit defaults (rather than user-passed params), because those parameters circumvent the need to comply with the RJSF-schema.

So I will leave these tests failing for the moment and go solve #373, and then come back here and see if the errors are resolved as expected.

**Pull request merged**